### DISCUSSION TITLE
plugin.json templating enhancement (routing to "https://{ClientParams.dynamicServerUrl}/some/path")

### DISCUSSION THREAD

**Discussion by dkozar** [grafana/grafana#38322] [Plugins]
**What would you like to be added**:

It would be great if templating in` plugin.json` could be extended to support calling arbitrary endpoints.
We're aware this is currently forbidden due to security, however in some cases this is needed.

For instance, currently one could use JsonData variables when specifiying the final URL (the proxy routes the traffic to):

`https://login.microsoftonline.com/{{ .JsonData.tenantId }}/oauth2/v2.0/token`

Note: `JsonData` is an object configured by the user via ConfigEditor:

![image](https://user-images.githubusercontent.com/1078403/94865301-4148cd00-0435-11eb-8009-72f07b7a6677.png)

(example `plugin.json` of Azure Monitor datasource [here](https://github.com/grafana/grafana/blob/7388a4fb743b8c46493cee3ecdcafda735554f2e/public/app/plugins/datasource/grafana-azure-monitor-datasource/plugin.json#L35))

What we would like to be added is a possibility of sending some dynamic variables from the browser to Grafana proxy route (i.e. not only the ones configured via ConfigEditor), for instance a random server endpoint (https://abc.some-domain.com, or https://def.some-domain.com) used like this:

`https://{ClientParams.serverUrl}/some/path`

Yes, we're aware of the security applications, but this would not be a default behavior and would need to be enabled in Grafana configuration.

**Why is this needed**:

Currently, the only solution to this would be a proxy that needs to somehow get the information of the final server to call (via request headers for instance) and then this proxy would route to this server. However, this is not elegant and requires additional resources.

Together with a [proxy in front of Grafana for resolving a TLS cypher order issue](https://github.com/grafana/grafana/issues/27984), this would mean running and maintaining 2 additional proxies, whereas this could possibly be solved by Grafana's own proxy.

Additional question: if we eventually implement this by modifying Grafana code, would this be considered as something that could be contributed to Grafana and accepted as a valid use case?

Thanks! 😄

**Comment by AllSeeingEye** [comment]
Up. This is a much-needed feature for the data source plugins.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
